src/popover/popover.ts

```diff
@@ -204,10 +233,49 @@ export class NgbPopover implements OnInit, OnDestroy, OnChanges {
               this._elementRef.nativeElement, this._windowRef.location.nativeElement, this.placement,
               this.container === 'body'));
 
+      if (this.autoClose) {
+        this._ngZone.runOutsideAngular(() => {
+          let justOpened = true;
+          requestAnimationFrame(() => justOpened = false);
+
+          const escapes$ = fromEvent<KeyboardEvent>(this._document, 'keyup')
+                               .pipe(takeUntil(this.hidden), filter(event => event.which === Key.Escape));
+
+          const clicks$ = fromEvent<MouseEvent>(this._document, 'click')
+                              .pipe(
+                                  takeUntil(this.hidden), filter(() => !justOpened),
+                                  filter(event => this._shouldCloseFromClick(event)));
+
+          race<Event>([escapes$, clicks$]).subscribe(() => this._ngZone.run(() => this.close()));
+        });
+      }
+
       this.shown.emit();
     }
   }
 
+  private _shouldCloseFromClick(event: MouseEvent) {
```

Comment:
could you please put private methods after all public methods?

Reply:
Will do